### PR TITLE
feat(spid): Increase TPM RDFIFO depth to 16

### PR DESCRIPTION
**Pull request by eunchan**
This commit increases the TPM Read FIFO size to 64 Bytes (16 x 4B).

The previous impl of TPM Read FIFO relies on the SW latency. The SW
latency should not be long so that SW is able to write the Read data
continuously. In order to satisfy the assumption, SW shall turn off
interrupts.

The requirement above could not comply with SW architecture. So, this
commit increases the size for SW to write full 64 B at once.

Latter design change will follow. The design change is for HW to send
the TPM Read data only when full payload size is ready in the FIFO.

### CODE REVIEW

**Comment by arnonsha** [comment]
@zi-v

**Comment by tjaychen** [comment]
@eunchan just to quickly double check. When you say `Latter design change will follow. The design change is for HW to send
the TPM Read data only when full payload size is ready in the FIFO.`, you don't mean we only send out once we have 64B right? But rather whenever the software decides the full payload is "ready"?

**Comment by eunchan** [comment]
> @eunchan just to quickly double check. When you say `Latter design change will follow. The design change is for HW to send the TPM Read data only when full payload size is ready in the FIFO.`, you don't mean we only send out once we have 64B right? But rather whenever the software decides the full payload is "ready"?
>
Right. The HW will check if the content in RdFIFO is greater than or equal to the TPM command header's xfer_size.

**Comment by a-will** [comment]
> @eunchan just to quickly double check. When you say `Latter design change will follow. The design change is for HW to send the TPM Read data only when full payload size is ready in the FIFO.`, you don't mean we only send out once we have 64B right? But rather whenever the software decides the full payload is "ready"?

I believe it'd be when the FIFO has the amount of data requested by the TPM master. If memory serves, the protocol has the master sending how much data it wants, and spi_device has just the one place to backpressure the master. So we'd fill the FIFO with the entire response before releasing from the wait state.

**Comment by tjaychen** [comment]
sounds good, thanks for explaining all.

**Review by tjaychen — APPROVED** [review]

**Review by weicaiyang — APPROVED** [review]
LGTM

File: sw/device/lib/dif/dif_spi_device.c [hunk 1]
```diff
@@ -1542,15 +1542,28 @@ dif_result_t dif_spi_device_tpm_write_data(dif_spi_device_handle_t *spi,
...
+  for (int i = 0; i < length; i += 4) {
+    if (i + 4 > length) {
+      // Send partial
+      rdfifo_wdata = 0;
+      for (int j = 0; j <= offset; j++) {
+        rdfifo_wdata |= buf[i + j] << (8 * j);
+      }
+    } else {
+      // Type casting to uint32_t then fetch
+      rdfifo_wdata = *((uint32_t *)buf + (i >> 2));
```
Comment (a-will) [inline sw/device/lib/dif/dif_spi_device.c:1563]: This is unfortunately not safe--We can't upcast to a wider size and use the pointer directly. Instead, probably need something like...
```suggestion
      // Assemble a 32-bit word.
      rdfifo_wdata = (buf[i+3] << 24) | (buf[i+2] << 16) | (buf[i+1] << 8) | buf[i];
```

This assumes the data is little-endian.

An alternative would be to specify the input array as being a uint32_t* and forcing the caller to handle alignment.

Comment (eunchan) [reply sw/device/lib/dif/dif_spi_device.c:1563]: I think the latter is better (using uint32_t*). Could you revise the DIF after this is merged?

**Pull request merged**